### PR TITLE
add train_pylate_contrastive.py

### PR DESCRIPTION
**Pull request by sigridjineth**
**Changes**
* [x] Add Example for training ColBERT using Pylate in terms of contrastive way. The existing script only shows the example for training with teacher distillation strategy.

### CODE REVIEW

**Comment by NohTow** [comment]
Hello,

I don't have my computer right now, but I also trained models using contrastive loss during the experiments for the paper and IIRC, the boilerplates from PyLate work out of the box if you just do not compile explicitely (as we compile internally as I specified in the other issue).

So I think I rather prefer we do that as I am not sure about the side effects the parameters of the inductor can have and we just don't need to compile explicitely (actually now the parameter to compile models in ST has been fixed, so even for the other models we should offload this using this parameter).

**Comment by sigridjineth** [comment]
@NohTow Yes, here, I used the Pylate boilerplate and worked pretty well out of the box when not explicitly compiling it. Have you any kinds of concerns not compiling the model when training as in my script?

**Comment by NohTow** [comment]
> Have you any kinds of concerns not compiling the model when training as in my script?

Not at all, especially as the ModernBERT  _is_ compiled by default (even if you do not call `model = torch.compile(model)`. My comment is rather to remove these two lines:
```
torch._inductor.config.fallback_random = True
torch._inductor.config.triton.unique_kernel_names = True
```
As they are not needed if you do not have the `model = torch.compile(model)`line and I am not sure about some side effects it could silently induce!

File: examples/train_pylate_contrastive.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+)
+
+from pylate import evaluation, losses, models, utils
+
+# Add at the start of your train_pylate_contrastive.py
+import torch
+torch._inductor.config.fallback_random = True
+torch._inductor.config.triton.unique_kernel_names = True
+# Or completely disable torch compile
+# model.forward = torch.compile(model.forward, mode="max-autotune", fullgraph=True)
```
Comment (NohTow) [inline examples/train_pylate_contrastive.py:15]:
```suggestion
# ModernBERT is compiled by default, so there is no need to call compile explicitly and it actually breaks the model
# model = torch.compile(model)
```

File: examples/train_pylate_contrastive.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    SentenceTransformerTrainingArguments,
+)
+
+from pylate import evaluation, losses, models, utils
+
+# Add at the start of your train_pylate_contrastive.py
+import torch
+torch._inductor.config.fallback_random = True
+torch._inductor.config.triton.unique_kernel_names = True
+# Or completely disable torch compile
```
Comment (NohTow) [inline examples/train_pylate_contrastive.py:14]:
```suggestion
```

File: examples/train_pylate_contrastive.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    SentenceTransformerTrainer,
+    SentenceTransformerTrainingArguments,
+)
+
+from pylate import evaluation, losses, models, utils
+
+# Add at the start of your train_pylate_contrastive.py
+import torch
+torch._inductor.config.fallback_random = True
+torch._inductor.config.triton.unique_kernel_names = True
```
Comment (NohTow) [inline examples/train_pylate_contrastive.py:13]:
```suggestion
```

File: examples/train_pylate_contrastive.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+from sentence_transformers import (
+    SentenceTransformerTrainer,
+    SentenceTransformerTrainingArguments,
+)
+
+from pylate import evaluation, losses, models, utils
+
+# Add at the start of your train_pylate_contrastive.py
+import torch
+torch._inductor.config.fallback_random = True
```
Comment (NohTow) [inline examples/train_pylate_contrastive.py:12]:
```suggestion
```

File: examples/train_pylate_contrastive.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+from datasets import load_dataset
+from sentence_transformers import (
+    SentenceTransformerTrainer,
+    SentenceTransformerTrainingArguments,
+)
+
+from pylate import evaluation, losses, models, utils
+
+# Add at the start of your train_pylate_contrastive.py
+import torch
```
Comment (NohTow) [inline examples/train_pylate_contrastive.py:11]:
```suggestion
```

File: examples/train_pylate_contrastive.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+from datasets import load_dataset
+from sentence_transformers import (
+    SentenceTransformerTrainer,
+    SentenceTransformerTrainingArguments,
+)
+
+from pylate import evaluation, losses, models, utils
+
+# Add at the start of your train_pylate_contrastive.py
```
Comment (NohTow) [inline examples/train_pylate_contrastive.py:10]:
```suggestion
```

File: examples/train_pylate_contrastive.py (hunk 1 above)
Comment (NohTow) [reply examples/train_pylate_contrastive.py:15]: So my suggestion is to remove everything altogether
If you want people to not try compiling the model and having the same issue as you, you can add this comment on top of the commented compile code to let them know.

**Comment by sigridjineth** [comment]
@NohTow okay, request it again

**Comment by NohTow** [comment]
I took the liberty to modify a bit the script to make it closer to the other boilerplate on this repository.
The script works so it can be merged, I just don't know about the batch size and learning rate (as I did not really explore much training ModernColBERT with contrastive learning), but I guess that is ok to not have optimal HP for boilerplates.
Although the LR seems very small to me compared to what we are used to with ModernBERT, did you run some sweeps?
